### ISSUE TITLE
Parsing remote GeoJSON input

### ISSUE THREAD

**Issue by rolego** [mapnik/mapnik#2647]
The project I'm currently working on has the need to be able to load GeoJSON data at runtime from an application server. Thus it would be very helpful if HTTP(S) would be supported as third source type beside local file and embedded.

What do you think?

**Comment by springmeyer** [comment]
No, it does not make sense for Mapnik itself to speak in http(s). Instead what you want to do is use a solid, existing http(s) client to download the data and then pass it along to mapnik for rendering. There are lots of ways to accomplish this in a very high performance and threadsafe manner but the design depends upon your goals. For now I'll just say that one design that has worked well is:

   - Pre-render any base tiles you might want to overlay the geojson on (or at least have separate mapnik.Maps and url requests to render them on the fly)
   - Download geojson in your application code, say node.js or python
   - Create a Mapnik XML document that uses the Mapnik `geojson` plugin and passes the geojson using the `inline` keyword which allows Mapnik to parse the geojson out of the XML string rather than a file
  - Load this XML doc from an in memory string into mapnik
  - render to a tile
  - composite that tile on top of another tile that might represent the basemap

This is how the Mapbox static API works (https://www.mapbox.com/developers/api/static/#overlay) which generates images very quickly like: https://api.tiles.mapbox.com/v4/examples.map-zr0njcqy/geojson(%7B%22type%22%3A%22Feature%22%2C%22properties%22%3A%7B%22stroke%22%3A%22%23f44%22%2C%22stroke-opacity%22%3A0.75%2C%22stroke-width%22%3A8%7D%2C%22geometry%22%3A%7B%22type%22%3A%22LineString%22%2C%22coordinates%22%3A%5B%5B-122.44227826595306%2C37.79718494691698%5D%2C%5B-122.44171768426895%2C37.79438511142195%5D%2C%5B-122.4367555975914%2C37.795016726275364%5D%2C%5B-122.4363800883293%2C37.79313670242186%5D%2C%5B-122.43802696466445%2C37.792924746337285%5D%5D%7D%7D)/-122.4418,37.7939,15/500x300.png?access_token=pk.eyJ1Ijoic3ByaW5nbWV5ZXIiLCJhIjoiSERsdHN5byJ9.CTjY8Y6NJ6cLW7cq7RDiVw

**Comment by rolego** [comment]
Thank your for your fast reply and detailed information!
You are right that http communication is out of scope of mapnik.

I did try the apporoach you suggested with the OGR plugin and was faced with the problem that the configuration xml got too big, was truncated and could not parsed anymore. The Geojson plugin could work and is worth a try. Is there a documentation or at least example how to configure this?

**Comment by springmeyer** [comment]
> faced with the problem that the configuration xml got too big, was truncated and could not be parsed anymore

That is odd. Could it have been a problem escaping data rather than being too big though? I'd be interested if you could replicate a testcase for what you saw.

How many MB's (if you serialized to disk) is the GeoJSON?

> The Geojson plugin might work and is worth a try. Is there a documentation or at least example how to configure this?

You just pass the string using the `inline` keyword (same a the `string` keyword you used for OGR plugin): https://github.com/mapnik/mapnik/blob/master/tests/visual_tests/styles/geojson-geometry-collection.xml#L17-L49

However, if you data is really large then serializing to a file first might actually be faster to render. To know how to advise more I'd need a better sense of you data size and what you are trying to accomplish.

**Comment by jbeuckm** [comment]
Can you link any example code with this approach? I need to send a query object to a remote server which responds with geojson. Then I need to render that into tiles. Is this inline/in-memory xml approach performant for doing this in realtime?
